thrust/thrust/system/cuda/detail/temporary_buffer.h

```diff
@@ -26,4 +26,107 @@
 #  pragma system_header
 #endif // no system header
 
-// this system has no special temporary buffer functions
+#include <thrust/detail/type_traits/pointer_traits.h>
+#include <thrust/system/cuda/detail/par.h>
+#include <thrust/system/detail/bad_alloc.h>
+
+THRUST_NAMESPACE_BEGIN
+namespace cuda_cub
+{
+
+// If par_nosync does not have a user provided allocator attached, these
+// overloads should be selected.
+
+template <typename T>
+_CCCL_HOST pair<T*, ::cuda::std::ptrdiff_t> get_temporary_buffer(par_nosync_t&, ::cuda::std::ptrdiff_t n)
+{
+  void* ptr;
+  cudaError_t status = cudaMallocAsync(&ptr, sizeof(T) * n, nullptr);
+
+  if (status != cudaSuccess)
+  {
+    cudaGetLastError(); // Clear the CUDA global error state.
+
+    // That didn't work. We could be somewhere where async allocation isn't
+    // supported like Windows, so try again with cudaMalloc.
+    status = cudaMalloc(&ptr, sizeof(T) * n);
+
+    if (status != cudaSuccess)
+    {
+      throw system::detail::bad_alloc(cuda_category().message(status).c_str());
+    }
+  }
```

Comment:
That should be a compile time switch and not runtime

Reply:
What do you mean? Status is a runtime variable.

Reply:
Ping - can you clarify?